src/app/developer/components/utils.ts

```diff
@@ -10,11 +45,39 @@ export async function getBooks(
   const books: Book[] = [];
 
   for (const [bookId, {defaultVersion}] of bookEntries) {
-    const bookLoader = archiveLoader.book(bookId, defaultVersion);
-    const osWebBook = await osWebLoader.getBookFromId(bookId);
-    const archiveBook = await bookLoader.load();
-
-    books.push(formatBookData(archiveBook, osWebBook));
+    books.push(await getBook(archiveLoader, osWebLoader, bookId, defaultVersion));
   }
   return books;
 }
+
+async function getBook(
+    archiveLoader: AppServices['archiveLoader'],
+    osWebLoader: AppServices['osWebLoader'],
+    bookId: string,
+    defaultVersion: string) {
+  const bookLoader = archiveLoader.book(bookId, defaultVersion);
+  const osWebBook = await osWebLoader.getBookFromId(bookId);
+  const archiveBook = await bookLoader.load();
+  return formatBookData(archiveBook, osWebBook);
+}
+
+export async function getCanonicalUrl(
+  archiveLoader: AppServices['archiveLoader'],
+  osWebLoader: AppServices['osWebLoader'],
+  bookId: string,
+  pageShortId: string
+) {
+  const canonicals: string[] = CANONICAL_MAP[bookId] || [];
+  for (const id of canonicals) {
+    const version = assertDefined(BOOKS[id], `Could not find ${id} in BOOKS config`).defaultVersion;
+
+    const canonicalBook = await getBook(archiveLoader, osWebLoader, id, version);
+    const pageInBook = getUrlParamForPageId(canonicalBook, pageShortId, true);
+
+    if (pageInBook) {
+      return {bookSlug: canonicalBook.slug, pageSlug: pageInBook};
```

Comment:
why are you passing around the page params instead of the ArchivePage object? there are helpers to construct urls that you can use when you do the urling